### PR TITLE
Imporove USB Debugging

### PR DESCRIPTION
**Pull request by cnlohr**
* Show "backtrace" even though only one item.
* Add note for where we can duplicate printf to uart.
* Add `installudev` rules
* Improve reboot-reflash timing.

### CODE REVIEW

File: makefile [hunk 1]
```diff
@@ -367,18 +367,29 @@ usbflash :
 	tools/reflash_and_monitor.bat
 else
 usbflash :
```
Comment (AEFeinstein) [inline makefile:369]: Tested on Windows too?

**Review by AEFeinstein — CHANGES_REQUESTED** [review]

File: makefile [hunk 2]
```diff
@@ -367,18 +367,29 @@ usbflash :
...
 monitor :
 	$(MAKE) -C tools/swadgeterm monitor
 
+installudev :
+	printf "KERNEL==\"hidraw*\", SUBSYSTEM==\"hidraw\", MODE=\"0664\", GROUP=\"plugdev\", ATTRS{idVendor}==\"1209\", ATTRS{idProduct}==\"4269\"\n" > /tmp/99-swadge.rules
+	printf "KERNEL==\"hidraw*\", SUBSYSTEM==\"hidraw\", ATTRS{idVendor}==\"1209\", ATTRS{idProduct}==\"4269\", GROUP=\"plugdev\", MODE=\"0660\"\n" >> /tmp/99-swadge.rules
+	printf "KERNEL==\"hidraw*\", SUBSYSTEM==\"hidraw\", MODE=\"0664\", GROUP=\"plugdev\", ATTRS{idVendor}==\"303a\", ATTRS{idProduct}==\"00??\"\n" >> /tmp/99-swadge.rules
+	printf "KERNEL==\"hidraw*\", SUBSYSTEM==\"hidraw\", ATTRS{idVendor}==\"303a\", ATTRS{idProduct}==\"00??\", GROUP=\"plugdev\", MODE=\"0660\"\n" >> /tmp/99-swadge.rules
+	sudo cp -a /tmp/99-swadge.rules /etc/udev/rules.d/99-swadge.rules
+	sudo usermod -aG plugdev cnlohr
```
[the remark this answers is not in the document]
Comment (dylwhich) [reply makefile:389]: And I don't even have a `plugdev` group!

[the remark this answers is not in the document]
Comment (cnlohr) [reply makefile:389]: bah

also @dylwhich how do you enable plugdev rules?

File: components/hdw-usb/advanced_usb_control.c [hunk 3]
```diff
@@ -84,6 +85,10 @@ static int advanced_usb_write_log(void* cookie __attribute__((unused)), const ch
         advanced_usb_printf_head                             = next;
         next                                                 = (advanced_usb_printf_head + 1) % AUPB_SIZE;
     }
+
+    // Write out to UART as well.
+    //uart_write_bytes(0, (const char*)data, size);
```
[the remark this answers is not in the document]
Comment (cnlohr) [reply components/hdw-usb/advanced_usb_control.c:90]: Correct.

File: docs/SERIAL_DEBUG.md [hunk 4]
```diff
@@ -111,7 +111,21 @@ This option will route serial debug output through a pin on the Simple Add-On (S
 
 ### DEBUG_OUTPUT_USB
 
-This option will route serial debug output over the USB header. The Swadge doesn't enumerate as a regular serial port, so you must run our own serial monitor, [`swadgeterm`](../tools/swadgeterm/) to view the output. To build `swadgeterm` from the root of this repository, run:
+This option will route serial debug output over the USB header. In general you can perform the following two steps to accomplish flashing and logging over the USB interface while the swadge is operating.
+
+This command will install udev rules on Linux to allow you to use the swadge for printf and other advanced USB features. **You only need to call it once**
```
[the remark this answers is not in the document]
Comment (cnlohr) [reply docs/SERIAL_DEBUG.md:116]: Dunno.  Will need to poke at it more. Probably not too bad?  We could also make it a script that adds it, then call that from a makefile?

File: makefile (hunk 1 above)
Comment (cnlohr) [reply makefile:369]: Will test later - this is not ready for merge.

**Comment by dylwhich** [comment]
This is what I get, after installing the rules (aside from the `plugdev` group which does not exist for Fedora).

```
$ make usbflash
# In case we are already in the bootloader...
(make -C tools/bootload_reboot_stub reboot)||(true)
make[1]: Entering directory '/home/dylwhich/esp/Swadge-IDF-5.0/tools/bootload_reboot_stub'
esptool.py --before no_reset --after no_reset load_ram bootload_reboot_stub.bin
esptool.py v4.7.0
Found 32 serial ports
Serial port /dev/ttyS9
/dev/ttyS9 failed to connect: Could not open /dev/ttyS9, the port is busy or doesn't exist.
(Could not configure port: (5, 'Input/output error'))
...
...
...
Serial port /dev/ttyS10
/dev/ttyS10 failed to connect: Could not open /dev/ttyS10, the port is busy or doesn't exist.
(Could not configure port: (5, 'Input/output error'))

Serial port /dev/ttyS1
/dev/ttyS1 failed to connect: Could not open /dev/ttyS1, the port is busy or doesn't exist.
(Could not configure port: (5, 'Input/output error'))

Serial port /dev/ttyS0
/dev/ttyS0 failed to connect: Could not open /dev/ttyS0, the port is busy or doesn't exist.
(Could not configure port: (5, 'Input/output error'))


A fatal error occurred: Could not connect to an Espressif device on any of the 32 available serial ports.
make[1]: *** [Makefile:4: reboot] Error 2
make[1]: Leaving directory '/home/dylwhich/esp/Swadge-IDF-5.0/tools/bootload_reboot_stub'
# Command reboot out of game into bootloader.
make -C tools/reboot_into_bootloader
make[1]: Entering directory '/home/dylwhich/esp/Swadge-IDF-5.0/tools/reboot_into_bootloader'
gcc -o reboot_swadge_into_bootloader reboot_swadge_into_bootloader.c -g -O0 -ludev
./reboot_swadge_into_bootloader || true
Could not open USB (1209:4269)
Could not open USB (1209:4269)
Could not open USB (1209:4269)
Could not open USB (1209:4269)
Could not open USB (1209:4269)
...
...
...
Could not open USB (1209:4269)
make[1]: Leaving directory '/home/dylwhich/esp/Swadge-IDF-5.0/tools/reboot_into_bootloader'
idf.py flash
Executing action: flash
Serial port /dev/ttyS9
/dev/ttyS9 failed to connect: Could not open /dev/ttyS9, the port is busy or doesn't exist.
(Could not configure port: (5, 'Input/output error'))

Serial port /dev/ttyS8
/dev/ttyS8 failed to connect: Could not open /dev/ttyS8, the port is busy or doesn't exist.
(Could not configure port: (5, 'Input/output error'))
...
...
...
Serial port /dev/ttyS0
/dev/ttyS0 failed to connect: Could not open /dev/ttyS0, the port is busy or doesn't exist.
(Could not configure port: (5, 'Input/output error'))

No serial ports found. Connect a device, or use '-p PORT' option to set a specific port.
make: *** [makefile:374: usbflash] Error 2
```

**Comment by dylwhich** [comment]
Okay, I got it working but I had to modify the udev rules as well to use my user group rather than `plugdev`

**Comment by dylwhich** [comment]
Also, my swadge was previously flashed with the old version that didn't have our PID, so I think I needed to flash it manually once.

File: makefile (hunk 2 above)
[the remark this answers is not in the document]
Comment (dylwhich) [reply makefile]: Apparently I just Don't Have Them, and it works fine? I just use my user's group instead and it seems to work fine. I pushed a change that checks if `plugdev` exists or not and uses your group instead.

**Comment by cnlohr** [comment]
> Okay, I got it working but I had to modify the udev rules as well to use my user group rather than `plugdev`

Should we add plugdev to groups?  there's no great solution I'm aware of.  `adm` `dialout` may be some of the others that would make sense.

File: docs/SERIAL_DEBUG.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (cnlohr) [reply docs/SERIAL_DEBUG.md:116]: Looks like dylwhich did it.

File: docs/SERIAL_DEBUG.md [hunk 5]
```diff
@@ -111,7 +111,21 @@ This option will route serial debug output through a pin on the Simple Add-On (S
...
+make installudev
+```
+
+Then any time you wish to reboot the swadge into bootloader, flash, reboot back into code and monitor, execute the following:
+```bash
+make usbflash
+```
+
+For more background,
+
```
[the remark this answers is not in the document]
Comment (cnlohr) [reply docs/SERIAL_DEBUG.md]: resolved

**Review by AEFeinstein — APPROVED** [review]

**Pull request merged**